Add nn.CircularPad{*}d for consistency + fix no_batch_dim support

mikaylagawarecki opened the pull request:
Fixes #105749 https://github.com/pytorch/pytorch/issues/95320 

(tldr is that input should always be `[N, C, H, (W, D])` where only H, W and D dimensions get circular padding, so the 2D case where user wants both dimensions to be padded --> they should `.unsqueeze(0)` (as is the case for `Reflection/ReplicationPad`) but we didn't document this for circular padding. [This seems to be the old docstring](https://github.com/pytorch/pytorch/blob/277b05014aced075337d6f84a535bd5150aabebb/torch/nn/functional.py#L4689) that was somehow lost.

Fixes no_batch_dim support https://github.com/pytorch/pytorch/issues/104860


- Adds missing documentation for circular padding
- Adds missing CircularPad modules
- Migrates legacy test_nn tests from circular padding to ModuleInfo
- Adds no_batch_dim support + sample inputs that test this

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106148
* #106147
* #106325



cc @albanD @mruberry @jbschlosser @walterddr

mikaylagawarecki committed:
Add nn.CircularPad{*}d for consistency + fix no_batch_dim support

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add nn.CircularPad{*}d for consistency + fix no_batch_dim support"

Fixes #105749 https://github.com/pytorch/pytorch/issues/95320 

(tldr is that input should always be `[N, C, H, (W, D])` where only H, W and D dimensions get circular padding, so the 2D case where user wants both dimensions to be padded --> they should `.unsqueeze(0)` (as is the case for `Reflection/ReplicationPad`) but we didn't document this for circular padding

Fixes no_batch_dim support https://github.com/pytorch/pytorch/issues/104860


- Adds missing documentation for circular padding
- Adds missing CircularPad modules
- Migrates legacy test_nn tests from circular padding to ModuleInfo
- Adds no_batch_dim support + sample inputs that test this




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add nn.CircularPad{*}d for consistency + fix no_batch_dim support"

Fixes #105749 https://github.com/pytorch/pytorch/issues/95320 

(tldr is that input should always be `[N, C, H, (W, D])` where only H, W and D dimensions get circular padding, so the 2D case where user wants both dimensions to be padded --> they should `.unsqueeze(0)` (as is the case for `Reflection/ReplicationPad`) but we didn't document this for circular padding

Fixes no_batch_dim support https://github.com/pytorch/pytorch/issues/104860


- Adds missing documentation for circular padding
- Adds missing CircularPad modules
- Migrates legacy test_nn tests from circular padding to ModuleInfo
- Adds no_batch_dim support + sample inputs that test this




[ghstack-poisoned]

albanD reviewed (approved):
Very nice update to the kernel and love to see more testing!

mikaylagawarecki committed:
Update on "Add nn.CircularPad{*}d for consistency + fix no_batch_dim support"

Fixes #105749 https://github.com/pytorch/pytorch/issues/95320 

(tldr is that input should always be `[N, C, H, (W, D])` where only H, W and D dimensions get circular padding, so the 2D case where user wants both dimensions to be padded --> they should `.unsqueeze(0)` (as is the case for `Reflection/ReplicationPad`) but we didn't document this for circular padding. [This seems to be the old docstring](https://github.com/pytorch/pytorch/blob/277b05014aced075337d6f84a535bd5150aabebb/torch/nn/functional.py#L4689) that was somehow lost.

Fixes no_batch_dim support https://github.com/pytorch/pytorch/issues/104860


- Adds missing documentation for circular padding
- Adds missing CircularPad modules
- Migrates legacy test_nn tests from circular padding to ModuleInfo
- Adds no_batch_dim support + sample inputs that test this




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add nn.CircularPad{*}d for consistency + fix no_batch_dim support"

Fixes #105749 https://github.com/pytorch/pytorch/issues/95320 

(tldr is that input should always be `[N, C, H, (W, D])` where only H, W and D dimensions get circular padding, so the 2D case where user wants both dimensions to be padded --> they should `.unsqueeze(0)` (as is the case for `Reflection/ReplicationPad`) but we didn't document this for circular padding. [This seems to be the old docstring](https://github.com/pytorch/pytorch/blob/277b05014aced075337d6f84a535bd5150aabebb/torch/nn/functional.py#L4689) that was somehow lost.

Fixes no_batch_dim support https://github.com/pytorch/pytorch/issues/104860


- Adds missing documentation for circular padding
- Adds missing CircularPad modules
- Migrates legacy test_nn tests from circular padding to ModuleInfo
- Adds no_batch_dim support + sample inputs that test this




[ghstack-poisoned]

albanD reviewed (approved):
SGTM

mikaylagawarecki committed:
Update on "Add nn.CircularPad{*}d for consistency + fix no_batch_dim support"

Fixes #105749 https://github.com/pytorch/pytorch/issues/95320 

(tldr is that input should always be `[N, C, H, (W, D])` where only H, W and D dimensions get circular padding, so the 2D case where user wants both dimensions to be padded --> they should `.unsqueeze(0)` (as is the case for `Reflection/ReplicationPad`) but we didn't document this for circular padding. [This seems to be the old docstring](https://github.com/pytorch/pytorch/blob/277b05014aced075337d6f84a535bd5150aabebb/torch/nn/functional.py#L4689) that was somehow lost.

Fixes no_batch_dim support https://github.com/pytorch/pytorch/issues/104860


- Adds missing documentation for circular padding
- Adds missing CircularPad modules
- Migrates legacy test_nn tests from circular padding to ModuleInfo
- Adds no_batch_dim support + sample inputs that test this




cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add nn.CircularPad{*}d for consistency + fix no_batch_dim support"

Fixes #105749 https://github.com/pytorch/pytorch/issues/95320 

(tldr is that input should always be `[N, C, H, (W, D])` where only H, W and D dimensions get circular padding, so the 2D case where user wants both dimensions to be padded --> they should `.unsqueeze(0)` (as is the case for `Reflection/ReplicationPad`) but we didn't document this for circular padding. [This seems to be the old docstring](https://github.com/pytorch/pytorch/blob/277b05014aced075337d6f84a535bd5150aabebb/torch/nn/functional.py#L4689) that was somehow lost.

Fixes no_batch_dim support https://github.com/pytorch/pytorch/issues/104860


- Adds missing documentation for circular padding
- Adds missing CircularPad modules
- Migrates legacy test_nn tests from circular padding to ModuleInfo
- Adds no_batch_dim support + sample inputs that test this




cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge

mikaylagawarecki committed:
Update on "Add nn.CircularPad{*}d for consistency + fix no_batch_dim support"

Fixes #105749 https://github.com/pytorch/pytorch/issues/95320 

(tldr is that input should always be `[N, C, H, (W, D])` where only H, W and D dimensions get circular padding, so the 2D case where user wants both dimensions to be padded --> they should `.unsqueeze(0)` (as is the case for `Reflection/ReplicationPad`) but we didn't document this for circular padding. [This seems to be the old docstring](https://github.com/pytorch/pytorch/blob/277b05014aced075337d6f84a535bd5150aabebb/torch/nn/functional.py#L4689) that was somehow lost.

Fixes no_batch_dim support https://github.com/pytorch/pytorch/issues/104860


- Adds missing documentation for circular padding
- Adds missing CircularPad modules
- Migrates legacy test_nn tests from circular padding to ModuleInfo
- Adds no_batch_dim support + sample inputs that test this




cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge

mikaylagawarecki committed:
Update on "Add nn.CircularPad{*}d for consistency + fix no_batch_dim support"

Fixes #105749 https://github.com/pytorch/pytorch/issues/95320 

(tldr is that input should always be `[N, C, H, (W, D])` where only H, W and D dimensions get circular padding, so the 2D case where user wants both dimensions to be padded --> they should `.unsqueeze(0)` (as is the case for `Reflection/ReplicationPad`) but we didn't document this for circular padding. [This seems to be the old docstring](https://github.com/pytorch/pytorch/blob/277b05014aced075337d6f84a535bd5150aabebb/torch/nn/functional.py#L4689) that was somehow lost.

Fixes no_batch_dim support https://github.com/pytorch/pytorch/issues/104860


- Adds missing documentation for circular padding
- Adds missing CircularPad modules
- Migrates legacy test_nn tests from circular padding to ModuleInfo
- Adds no_batch_dim support + sample inputs that test this




cc albanD mruberry jbschlosser walterddr

[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge

mikaylagawarecki commented:
@pytorchbot merge

malfet commented:
@pytorchbot revert -m "Reverting as dependent PR https://github.com/pytorch/pytorch/pull/106147 was reverted as well" -c weird
